src/generic/complex_matrices.cc

```diff
@@ -538,6 +539,650 @@ namespace oomph
+  {
+#ifdef PARANOID
+    // check that this matrix is built
+    if (!Built)
+    {
+      std::ostringstream error_message_stream;
+      error_message_stream << "This matrix has not been built";
+      throw OomphLibError(error_message_stream.str(),
+                          OOMPH_CURRENT_FUNCTION,
+                          OOMPH_EXCEPTION_LOCATION);
+    }
+    // check that this matrix is built
+    if (!matrix_in.built())
+    {
+      std::ostringstream error_message_stream;
+      error_message_stream << "This matrix matrix_in has not been built";
+      throw OomphLibError(error_message_stream.str(),
+                          OOMPH_CURRENT_FUNCTION,
+                          OOMPH_EXCEPTION_LOCATION);
+    }
+    // if soln is setup then it should have the same distribution as x
+    if (result.built())
+    {
+      if (!(*result.distribution_pt() == *this->distribution_pt()))
+      {
+        std::ostringstream error_message_stream;
+        error_message_stream
+          << "The matrix result is setup and therefore must have the same "
+          << "distribution as the vector x";
+        throw OomphLibError(error_message_stream.str(),
+                            OOMPH_CURRENT_FUNCTION,
+                            OOMPH_EXCEPTION_LOCATION);
+      }
+    }
+#endif
+
+    // short name for Serial_matrix_matrix_multiply_method
+    unsigned method = 2;
+
+    // if this matrix is not distributed and matrix in is not distributed
+    if ((method == 1) || (method == 2) || (method == 3))
+    {
+      // NB N is number of rows!
+      unsigned long N = this->nrow();
+      unsigned long M = matrix_in.ncol();
+      unsigned long Nnz = 0;
+
+      // pointers to arrays which store result
+      int* Row_start = 0;
+      std::complex<double>* Value = 0;
+      int* Column_index = 0;
+
+      // get pointers to matrix_in
+      const int* matrix_in_row_start = matrix_in.row_start();
+      const int* matrix_in_column_index = matrix_in.column_index();
+      const std::complex<double>* matrix_in_value = matrix_in.value();
+
+      // get pointers to this matrix
+      const std::complex<double>* i_value = this->value();
+      const int* i_row_start = this->row_start();
+      const int* i_column_index = this->column_index();
+
+      // clock_t clock1 = clock();
+
+      // METHOD 1
+      // --------
+      if (method == 1)
+      {
+        // allocate storage for row starts
+        Row_start = new int[N + 1];
+        Row_start[0] = 0;
+
+        // a set to store number of non-zero columns in each row of result
+        std::set<unsigned> columns;
+
+        // run through rows of this matrix and matrix_in to find number of
+        // non-zero entries in each row of result
+        for (unsigned long i_row = 0; i_row < N; i_row++)
+        {
+          // run through non-zeros in i_row of this matrix
+          for (int i_non_zero = i_row_start[i_row];
+               i_non_zero < i_row_start[i_row + 1];
```

Comment:
Good coding practice to store `i_row_start[i_row + 1]` in a local variable outside the for-loop to avoid repeated memory accesses

Reply:
Fixed else where as well. (Definitions inside the outer loop though...)